rfcs/20190506-filesystem-plugin-modular-tensorflow.md

```diff
@@ -0,0 +1,1082 @@
+without any changes to the code.
+
+A user having a custom filesystem implementation can use it from Python by
+calling [`load_library`][load_library]. When the shared library is loaded, the
+plugin initialization starts, as mentioned before. Once the filesystem
+implementation is available to TensorFlow core, it can also be accessed from
+Python.
+
+## Versioning
+
+Since the code of a plugin can develop at a different velocity than the code of
+core TensorFlow, we need to ensure backwards and forwards compatibility. Any
+change in a method’s signature or reordering the methods in the `TF_*Ops` data
+structures needs a new version as plugins compiled against one signature will
+either crash or not load properly if core has a different interface.
+
+Thus, the `TF_*Ops` data structures need to record version information which is
+checked at plugin loading time. For example, this is a potential implementation
+of `TF_RandomAccessFileOps`, including versioning fields:
+
+```cpp
+typedef struct TF_RandomAccessFileOps {
+  // these fields must always be at the beginning
+  const int version = 1;
+  // alternatively: const int size = sizeof(TF_RandomAccessFileOps);
+  const int num_ops = 2;
+
+  // the real ops in the table
+  size_t (*const Read)(const TF_RandomAccessFile*, uint64, size_t, char*, TF_Status*);
+  void (*const Cleanup)(TF_RandomAccessFile*);
+} TF_RandomAccessFileOps;
+```
+
+We will allow each structure to evolve independently. That is, the `version`
+field in each one can be increased independently of the value of the field in
+the other structures.
+
+To increase the range of compatibility, we will give a guarantee that, except on
+new major versions, new methods will only be added to the end of each of these
+structures and that no method will be deleted from the table. Changing the
+signature of a method will be handled by adding the method with the new
+signature to the end of the table and marking the old method as deprecated.
+
+Deprecation of a method will be done in two parts. First, we will add compiler
+pragmas to warn the users compiling against the interface and relying on
+deprecated methods. Second, at plugin’s registration, TensorFlow code can check
+the function pointers sent by the plugin to see which ones are for deprecated
+methods. In case there is any, a warning message should be printed.
+
+To allow both the core and the plugin to know which methods from the tables are
+supported, we also add the `num_ops` field, counting how many methods are
+defined in the table. This field can be used to determine the amount of
+available API calls, as it is guaranteed to only increase during minor releases.
+We provide a `NUM_OPS_IN_TABLE` macro to allow the compiler to set the value of
+`num_ops` automatically.
+
+Alternatively, instead of `num_ops` field we can have a size field which is the
+size of the entire structure and can be obtained via `sizeof`. The benefits of
```

Comment:
The size field makes it easier to be memory-safe, if you do choose to go with exposing the struct to users.

Reply:
I agree. Added this to the questions too but I totally agree now.